beacon-chain/endtoend/endtoend_test.go

```diff
@@ -0,0 +1,514 @@
+	beaconNodes := startBeaconNodes(t, config)
+	valClients := initializeValidators(t, config, keystorePath, beaconNodes)
+
+	beaconLogFile, err := os.Open(path.Join(tmpPath, "beacon-0.log"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := WaitForTextInFile(beaconLogFile, "Chain started within the last epoch"); err != nil {
+		t.Fatal(err)
+	}
+	t.Log("Chain has started")
+
+	if config.numBeaconNodes > 1 {
+		t.Run("peers_connect", func(t *testing.T) {
+			for _, bNode := range beaconNodes {
+				if err := PeersConnect(bNode.monitorPort, config.numBeaconNodes-1); err != nil {
+					t.Fatalf("failed to connect to peers: %v", err)
+				}
+			}
+		})
+	}
+
+	conn, err := grpc.Dial("127.0.0.1:4000", grpc.WithInsecure())
+	if err != nil {
+		t.Fatalf("fail to dial: %v", err)
+	}
+	time.Sleep(2 * time.Second)
+	beaconClient := eth.NewBeaconChainClient(conn)
+
+	currentEpoch := uint64(0)
+	for currentEpoch < config.epochsToRun {
+		if currentEpoch > 0 {
+			newEpochText := fmt.Sprintf("\"Starting next epoch\" epoch=%d", currentEpoch)
+			if err := WaitForTextInFile(beaconLogFile, newEpochText); err != nil {
+				t.Fatal(err)
+			}
+		}
+
+		t.Logf("Current Epoch: %d\n", currentEpoch)
+		for _, evaluator := range config.evaluators {
+			t.Run(evaluator.Name, func(t *testing.T) {
+				if err := evaluator.Evaluation(beaconClient); err != nil {
+					t.Fatal(err)
+				}
+			})
+		}
+		currentEpoch++
+		time.Sleep(time.Second)
+	}
+
+	if currentEpoch < config.epochsToRun {
+		t.Fatalf("test ended prematurely, only reached epoch %d", currentEpoch)
+	}
+
+	processIDs := []int{eth1PID}
+	for _, vv := range valClients {
+		processIDs = append(processIDs, vv.processID)
+	}
+	for _, bb := range beaconNodes {
+		processIDs = append(processIDs, bb.processID)
+	}
+	for _, id := range processIDs {
+		process, err := os.FindProcess(id)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if err := process.Kill(); err != nil {
+			t.Fatalf("could not kill process: %v", err)
+		}
+	}
+}
+
+// startEth1 starts an eth1 local dev chain and deploys a deposit contract.
+func startEth1(t *testing.T, tmpPath string) (common.Address, string, int) {
+	binaryPath, found := bazel.FindBinary("cmd/geth", "geth")
+	if !found {
+		t.Fatal("go-ethereum binary not found")
+	}
+
+	args := []string{
+		fmt.Sprintf("--datadir=%s", path.Join(tmpPath, "eth1data/")),
+		"--rpc",
+		"--rpcaddr=0.0.0.0",
+		"--rpccorsdomain=\"*\"",
+		"--rpcvhosts=\"*\"",
+		"--ws",
+		"--wsaddr=0.0.0.0",
+		"--wsorigins=\"*\"",
+		"--dev",
+		"--dev.period=0",
+	}
+	cmd := exec.Command(binaryPath, args...)
+	file, err := os.Create(path.Join(tmpPath, "eth1.log"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	cmd.Stdout = file
+	cmd.Stderr = file
+	if err := cmd.Start(); err != nil {
+		t.Fatalf("failed to start eth1 chain: %v", err)
+	}
+
+	if err = WaitForTextInFile(file, "IPC endpoint opened"); err != nil {
+		t.Fatal(err)
+	}
+	// Small sleep to make sure IPC endpoint starts.
+	time.Sleep(4 * time.Second)
```

Comment:
Similar to my other comment, 4 seconds is quite arbitrary and can lead to flakes. Maybe it takes 4.1s on buildkite. Is there any way you can improve this? Maybe an exponential back off dialing IPC endpoint strategy with a deadline / timeout.

Reply:
Things were flaky at first so I was using this but removing all the random `time.Sleeps` doesn't cause any issues. :+1: